Storing metrics in the database

jesseobrien opened the pull request.

jesseobrien committed:
Storing metrics in the database

jesseobrien committed:
Implementing HTTP test with memory store

awly commented on go.mod:
```diff
@@ -3,10 +3,10 @@ module github.com/jesseobrien/fake-iot
 
 require (
-	github.com/davecgh/go-spew v1.1.1 // indirect
+	github.com/jackc/pgx/v4 v4.10.1
```
looking at `go.sum`, this added a ton of imports (about 100k lines in total)
consider using github.com/lib/pq to avoid that instead

jesseobrien replied:
Great point. I've used `pgx` before on a project but I hadn't paid any attention to the dependencies it creates. `lib/pq` for this seems like a better choice.

awly commented on internal/http/metrics.go:
```diff
@@ -3,33 +3,55 @@ package http
+		}
+
+		log.Printf("count of logins: %d", count)
+		// @TODO remove the above
```
per TODOs, remove this

awly commented on internal/http/metrics_test.go:
```diff
@@ -0,0 +1,45 @@
+
+func TestIngestMetrics(t *testing.T) {
+
+	t.Run("test that metrics will be written properly to the database", func(t *testing.T) {
```
nit: unnecessary subtest wrapper
use subtests when you have multiple of them.

jesseobrien replied:
I actually didn't mean to commit this with the `t.Run()` in it. I got a bit ahead of myself and then remembered that I don't need to write a stack of tests for every path so I just left this one. :sweat:

awly commented on internal/http/metrics_test.go:
```diff
@@ -0,0 +1,45 @@
+		rec := httptest.NewRecorder()
+		echoCtx := e.NewContext(req, rec)
+
+		if assert.NoError(t, handler(echoCtx)) {
```
nit: use `require.NoError` instead to avoid indentation

awly commented on internal/storage/metrics_mem.go:
```diff
@@ -0,0 +1,59 @@
+
+func (mas *MemAccountStore) WroteMetric(accountId, userId string) (bool, error) {
+
+	metrics, ok := mas.metrics[accountId]
```
lock the mutext before reading metrics

awly commented on internal/storage/metrics_mem.go:
```diff
@@ -0,0 +1,59 @@
+	mas.mu.Lock()
+	metrics := mas.metrics[metric.AccountID]
+	mas.metrics[metric.AccountID] = append(metrics, metric)
+	mas.mu.Unlock()
```
mutex should be locked for reading too

awly commented on internal/storage/metrics_mem.go:
```diff
@@ -0,0 +1,59 @@
+		return 0, fmt.Errorf("could not find metrics for account id %s", accountId)
+	}
+
+	return len(mas.metrics[accountId]), nil
```
lock the mutext before reading metrics

awly commented on internal/storage/metrics_mem.go:
```diff
@@ -0,0 +1,59 @@
+
+func (mas *MemAccountStore) CountByAccountId(ctx context.Context, accountId string) (int, error) {
+	if _, ok := mas.metrics[accountId]; !ok {
+		return 0, fmt.Errorf("could not find metrics for account id %s", accountId)
```
nit: not sure if this needs to be an error
if there are no metrics for an account, I'd expect `CountByAccountId` to return 0.
an error would mean something like "failed talking to the database".

jesseobrien replied:
I agree. It would make more sense to just return the count for the memory implementation as it can't "fail".

awly commented on internal/storage/metrics_pg.go:
```diff
@@ -0,0 +1,60 @@
+
+	sql := `INSERT INTO account_logins (user_id, account_id, timestamp) VALUES ($1, $2, $3)`
+
+	_, err = tx.Exec(ctx, sql, metric.UserID, metric.AccountID, metric.Timestamp)
```
this doesn't need to be in a transaction, since you're only doing a single Exec.

jesseobrien replied:
Fair point. I reworked it when switching to `lib/pq`.

jesseobrien committed:
Removing pgx in favour of lib/pq for postgres

jesseobrien committed:
Refactoring for PR comments

- Fixing mutex locks on reads in the memory implementation of the metric
store
- Other cleanup

russjones commented on Makefile:
```diff
@@ -15,12 +15,32 @@ build:
 
+.PHONY: test
+test:
+	go test ./...
```
Might want to add the data race detector in here as well.
```
go test -race ./...
```

russjones commented on internal/storage/metrics_mem.go:
```diff
@@ -0,0 +1,66 @@
+	mas.mu.Unlock()
+
+	return nil
+}
```
Why is this under two locks (also function always appears to return nil, so drop the error)? Why not the following:

```go
func (mas *MemAccountStore) Write(ctx context.Context, metric UserLoginMetric) {
	mas.mu.Lock()
	defer mas.mu.Unlock()

	if _, ok := mas.metrics[metric.AccountID]; !ok {
		mas.metrics[metric.AccountID] = []UserLoginMetric{metric}
		mas.mu.Unlock()
	}
	metrics := mas.metrics[metric.AccountID]
	mas.metrics[metric.AccountID] = append(metrics, metric)
}

jesseobrien replied:
I agree there should only be one lock. The error is part of the [interface for write](https://github.com/JesseObrien/fake-iot/blob/store-metrics/internal/storage/metrics.go#L16) so I don't necessarily want to change the method signature.

russjones replied:
Sure, if you feel you'll be returning errors at some point that's fine.

russjones commented on internal/storage/metrics_mem.go:
```diff
@@ -0,0 +1,66 @@
+	return nil
+}
+
+func (mas *MemAccountStore) CountByAccountId(ctx context.Context, accountId string) (int, error) {
```
Function doesn't appear to return an error ever.

jesseobrien replied:
Correct, however the [interface for the func](https://github.com/JesseObrien/fake-iot/blob/store-metrics/internal/storage/metrics.go#L17) has the `(int, error)` on it.

jesseobrien committed:
Add -race to `make watch` and fix locks on memory AccountStore

awly commented on internal/storage/metrics_mem.go:
```diff
@@ -0,0 +1,61 @@
+		return nil
+	}
+
+	metrics := mas.metrics[metric.AccountID]
```
`append` will work when `metrics` is nil
```suggestion
	metrics := mas.metrics[metric.AccountID]
```

awly commented on internal/storage/metrics_mem.go:
```diff
@@ -0,0 +1,61 @@
+func (mas *MemAccountStore) WroteMetric(accountId, userId string) (bool, error) {
+	mas.mu.Lock()
+	metrics, ok := mas.metrics[accountId]
+	mas.mu.Unlock()
```
```suggestion
	mas.mu.Lock()
	defer mas.mu.Unlock()
	metrics, ok := mas.metrics[accountId]
```

awly commented on internal/storage/metrics_mem.go:
```diff
@@ -0,0 +1,61 @@
+	if _, ok := mas.metrics[accountId]; !ok {
+		return 0, nil
+	}
+
```
`len` works on nil slices
```suggestion
```

jesseobrien committed:
Addressing locks and length checks
